domain/secretbackend/state/state.go

```diff
@@ -0,0 +1,321 @@
+		}
+		if err := tx.Query(ctx, rotationStmt, arg).Run(); err != nil {
+			return fmt.Errorf("deleting secret backend rotation for %q: %w", backendID, err)
+		}
+		if err = tx.Query(ctx, modelMetadataStmt, arg).Run(); err != nil {
+			return fmt.Errorf("resetting secret backend %q to NULL for model metadata: %w", backendID, err)
+		}
+		if err = tx.Query(ctx, backendStmt, arg).Run(); err != nil {
+			return fmt.Errorf("deleting secret backend for %q: %w", backendID, err)
+		}
+		return nil
+	})
+	return domain.CoerceError(err)
+}
+
+// ListSecretBackends returns a list of all secret backends.
+func (s *State) ListSecretBackends(ctx context.Context) ([]*coresecrets.SecretBackend, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	stmt, err := sqlair.Prepare(`
+SELECT 
+    b.uuid                  AS &SecretBackendRow.uuid,
+    b.name                  AS &SecretBackendRow.name,
+    b.backend_type          AS &SecretBackendRow.backend_type,
+    b.token_rotate_interval AS &SecretBackendRow.token_rotate_interval,
+    c.name                  AS &SecretBackendRow.config_name,
+    c.content               AS &SecretBackendRow.config_content
+FROM secret_backend b
+    LEFT JOIN secret_backend_config c ON b.uuid = c.backend_uuid
+ORDER BY b.name`, SecretBackendRow{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var rows SecretBackendRows
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt).GetAll(&rows)
+		if err != nil {
+			return fmt.Errorf("querying secret backends: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot list secret backends: %w", err)
+	}
+	return rows.toSecretBackends(), errors.Trace(err)
+}
+
+func (s *State) getSecretBackend(ctx context.Context, columName string, v string) (*coresecrets.SecretBackend, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	q := fmt.Sprintf(`
+SELECT
+    b.uuid                  AS &SecretBackendRow.uuid,
+    b.name                  AS &SecretBackendRow.name,
+    b.backend_type          AS &SecretBackendRow.backend_type,
+    b.token_rotate_interval AS &SecretBackendRow.token_rotate_interval,
+    c.name                  AS &SecretBackendRow.config_name,
+    c.content               AS &SecretBackendRow.config_content
+FROM secret_backend b
+    LEFT JOIN secret_backend_config c ON b.uuid = c.backend_uuid
+WHERE b.%s = $M.identifier`, columName)
+	stmt, err := sqlair.Prepare(q, sqlair.M{}, SecretBackendRow{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var rows SecretBackendRows
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, sqlair.M{"identifier": v}).GetAll(&rows)
+		if err != nil {
+			return fmt.Errorf("querying secret backends: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot list secret backends: %w", err)
+	}
```

Comment:
FYI: With the latest update of sqlair, the GetAll will return `sql.ErrNoRows` which means your len check below will fail.

Reply:
Oh, it's good to know this.